R/mixture_model_wrapper.R

```diff
@@ -0,0 +1,141 @@
+#' `function(...){EpiSoon::bsts_model(model = , ...)}`.
+#' @inheritParams bsts_model
+#' @return A dataframe of predictions (with columns representing the
+#'  time horizon and rows representing samples).
+#' @export
+#' @importFrom tsibble tsibble
+#' @importFrom fabletools model forecast
+#' @importFrom purrr map
+#' @importFrom dplyr bind_rows
+#' @examples
+#'
+#' y <- EpiSoon::example_obs_rts[1:10, ]$rt
+#' model = fable::ARIMA(y ~ time)
+#' samples = 10
+#' horizon = 7
+#' weighting_period = 5
+#'
+#' # make list with models
+#' baseline_ensemble <- function(...) {
+#' EpiSoon::fable_model(model = fabletools::combination_model(fable::ARIMA(y), fable::ETS(y), fable::NAIVE(y),
+#'                                                            fable::RW(y ~ drift()), cmbn_args = list(weights = "inv_var")), ...)
+#' }
+#' models <- list(
+#'   Mixfabl = baseline_ensemble,
+#'   "ARIMA" = function(...){EpiSoon::fable_model(model = fable::ARIMA(y), ...)},
+#'   "ETS" = function(...){EpiSoon::fable_model(model = fable::ETS(y), ...)},
+#'   "Drift" = function(...){EpiSoon::fable_model(model = fable::RW(y ~ drift()), ...)}
+#' )
+#'
+#' # make forecst
+#' forecast <- crps_ensemble(y = y,
+#'                           models = models,
+#'                           samples = 10,
+#'                           horizon = 7,
+#'                           weighting_period = 5)
+#'
+#'
+#'
+#' forecast_rt(EpiSoon::example_obs_rts[1:10, ],
+#'            model = function(...){
+#'              crps_ensemble(y = y,
+#'                            models = models,
+#'                            samples = 10,
+#'                            horizon = 7,
+#'                            weighting_period = 5)
+#'                          })
+#'
+#'
+
+
+crps_ensemble <- function(y = NULL,
+                          models = NULL,
+                          weighting_period = 5,
+                          ...) {
+
+
+  #### Error Handling
+  # check if stackr is installed
+  if (!suppressWarnings(require("stackr", quietly = TRUE) == TRUE)) {
+    stop("package stackr must be installed. You can install it using devtools::install_github('nikosbosse/stackr')")
+  }
+  if (length(y) <= weighting_period) {
+    stop("not enough observations to do weighting. Adjust weighting_period")
+  }
+
+  #### split data into train data and data for weighting. Use train data to
+  # generate forecasts, score them against weight data, then generate forecasts
+  # bases on the entire time series and use the obtained weights to create mixture
+  n <- length(y)
+  y_train <- y[1:(n - weighting_period)]
+  y_weight <- y[(n - weighting_period + 1):n]
+
+  #### fit models on train data and generate forecasts
+  fc_w <- lapply(seq_along(models),
+         FUN = function(i) {
+           f <- models[[i]]
+           out <- f(y = y_train, samples = samples, horizon = weighting_period)
+
+           # bring data in the correct format the stackr package expects
+           out %>%
```

Comment:
reduce pipe usage here as each pipe is a data copy

Reply:
reduced, but still a lot. data.table inside stackr looks most promising to me